WYSIWYG GUI for QSD Editing

willgearty opened the pull request:
This adds a very nice GUI to the QSD editing components of the website (including editable text on template pages and the .edit.html pages for static pages). I made sure that there is still a way to edit the source code directly for those that are familiar with HTML/markdown. Information about the GUI is [here](https://xdsoft.net/jodit/). I included almost all of the buttons that were offered, but let me know if any should be removed.

Fixes #2652.

Examples:

![image](https://user-images.githubusercontent.com/7232514/52248798-f9128980-28a4-11e9-955e-4348353788d6.png)

![image](https://user-images.githubusercontent.com/7232514/52248792-f021b800-28a4-11e9-9386-83340a546d24.png)

![image](https://user-images.githubusercontent.com/7232514/52248674-64a82700-28a4-11e9-9610-41661b550edb.png)

kkbrum commented:
I love this very much. I did not succeed in breaking anything, and the code looks good, so I think this is set!

Couple observations:
- the "add file" button doesn't make much sense to me, since you have to add the URL of a file and then it just links that URL. So it is essentially the same as the "add link" button, so can probably be removed?
- eventually (not on this PR probably), it would be nice to have the "add image" button accept uploads from the local computer and upload them to the filebrowser for you and display them. The adding image from URL is still a great feature though and i imagine coding up something to add the image from the computer to the filebrowser through this gui is a lot of work since it is not standard.

willgearty commented:
Actually, now that I'm looking around, it looks like there is the capability to upload images. It looks like there are two separate options:

1. Upload an image as a Base64 image, and the data for the image is stored in the HTML (see the option discussed in [this issue](https://github.com/jodit/jodit-angular/issues/9)). This is plug-and-play, although apparently there might be [problems](https://medium.com/snapp-mobile/dont-use-base64-encoded-images-on-mobile-13ddeac89d7c) using base64 on mobile?
2. Upload an image to the server using php (?), with documentation and examples [here](https://xdsoft.net/jodit/doc/types/IUploaderOptions/) and a tutorial [here](https://xdsoft.net/jodit/v.2/doc/tutorial-uploader-settings.html). It also looks like you can use similar methods to implement a [filebrowser](https://xdsoft.net/jodit/v.2/doc/tutorial-filebrowser-options.html) for the server-hosted images?

@benjaminjkraft As an actual developer, are you familiar with either of these things?

Either way, if we want to get this merged and pushed to chapters as part of the SR10-fixes, we can probably implement the image upload at a later time.

kkbrum commented:
I'm happy with this solution, but one thing I noticed is that if I now look at the source code after uploading an image, the image takes up a tonnnnnnnnn of lines in the source code so it might be pretty annoying if someone wanted to work with the html after uploading something. But if they want to use source code, they could just go through the filebrowser to avoid having annoying image info there...

willgearty commented:
Maybe base64 isn't the way to go, then... It looks like there are lots of [problems](https://stackoverflow.com/questions/17288290/why-not-embed-images-as-base64-strings-in-wygiwys-editors) with it, including the fact that the images are actually larger than the original files, they can slow down loading the page, and they can't be cached (and now what @kkbrum mentioned in the source code editor). But it's really easy to include in this PR, and I think we should get this change to chapters as soon as possible, so either we remove the base64 entirely now and then implement the actual upload to the server in a later PR or keep it for now and then implement the server upload. Either way, it'll take me a bunch of work to implement the server upload stuff.

willgearty commented:
This seems useful: https://www.froala.com/wysiwyg-editor/docs/server/django/file-upload

kkbrum commented:
I vote for removing the base64 and merging the rest of this gui because the rest is fabulous! The base64 was cool last time when it worked but this time I keep getting errors when it tries to save my qsd pages:
![image](https://user-images.githubusercontent.com/25392075/52984788-ed53b680-33be-11e9-8ee9-fc7b99ddaf61.png)

willgearty commented:
Yeah, I agree. After looking into it, we're storing so much data in the QSD database objects, that it's causing some serious memory issues. I've removed the base64 capability. I've implemented a temporary solution for the time being (link goes to the filebrowser):

![image](https://user-images.githubusercontent.com/7232514/52996958-f1d69a00-33d3-11e9-8775-d17c22299098.png)

kkbrum commented:
I like it!
